Re-add missing default spawn lists in features and use of insideOnly

sciwhiz12 opened the pull request:
This PR fixes #8265 by re-adding the missing default spawn lists (returned via `IForgeStructureFeature#getDefaultSpawnList`) to the four vanilla features of nether fortresses, ocean monuments, pillager outposts, and swampland huts.

The patches to the features for [nether fortresses][nether-fortress-1.18-deletion], [pillager outposts][pillager-outputs-1.18-deletion], and [swampland huts][swampland-hut-1.18-deletion] were removed in the 1.18 patch and update cycle.[^1] 

The patch for the [ocean monument][ocean-monument-1.18-change] remained, but it was slightly errorneous in that it removed an existing method in the `OceanMonumentFeature` class -- the `linearSeparation` method -- and replacing it with the `getDefaultSpawnList` method, rather than merely adding the new method.

Testing this PR with the steps in the linked issue (visting a pillager outpost and watching for pillager spawns) shows this fixes the issue.

---

This PR also fixes #8301 by modifying `StructureSpawnManager` to respect `insideOnly`, which controls whether the structure spawns occur within the structure's overall bounds or only within the 'inside' space of the structure (the bounds of each piece within the structure). 

This seems to have been introduced in the 1.18 patch and update cycle. The `StructureFeatureManager#getStructureAt(BlockPos, boolean, StructureFeature<?>)` method with the `insideOnly` boolean parameter was split into two: `getStructureAt` for the overall bounds of the structure feature, and `getStructureWithPieceAt` for the 'inside' space of the structure (the bounds of each piece). During patching, the second parameter was likely removed to fix the compilation error but was not revisited later to check the correctness of the code. <sup>#BlameCurle</sup>

[swampland-hut-1.18-deletion]: https://github.com/MinecraftForge/MinecraftForge/commit/40a174310d9eaf31e089f18b66a26d45c96d5587#diff-8c3a104cde9267723d45c08ddb2afe5fd1ac47dda3c48b008d7da44ba13903a9
[pillager-outputs-1.18-deletion]: https://github.com/MinecraftForge/MinecraftForge/commit/40a174310d9eaf31e089f18b66a26d45c96d5587#diff-37267cec0c0e9d464e54413c218521ac8da2bf733e1ecefc979c8d70e261e8e6
[nether-fortress-1.18-deletion]: https://github.com/MinecraftForge/MinecraftForge/commit/40a174310d9eaf31e089f18b66a26d45c96d5587#diff-f8480fb199973b9d8605133fca1201edaf4768fbba13150545c9ca69c68f7962
[ocean-monument-1.18-change]: https://github.com/MinecraftForge/MinecraftForge/commit/40a174310d9eaf31e089f18b66a26d45c96d5587#diff-f84183102ec66011846e5dac2404cd95cdee733fa2f7250893a7c7b9d01a3be1

[^1]: I presume these patches were removed because the method they used to remove as part of the patch was removed, and it was assumed that the patch was therefore unnecessary.<sup>#BlameCurle</sup>

sciwhiz12 committed:
Re-add missing default spawn lists in features

These were removed in 1.18, which meant no spawns occur in the following
features: nether fortresses, ocean monuments, pillager outposts, and
swampland huts.

For the ocean monument patch, this correct an oversight where an
existing method was removed (linearSeparation) and replaced with the
patched-in default spawn lists method.

Fixes #8265

pupnewfster commented:
It looks like there is also a missing patch in `NaturalSpawner#mobsAt` that should be redirecting `NetherFortressFeature.FORTRESS_ENEMIES` to `StructureSpawnManager#getStructureSpawns`, except if it is null (say a mod wants to remove all fortress spawns) then it will need to pass an empty list instead of null.

sciwhiz12 committed:
Patch in #getSpawnList for NETHER_BRIDGE

This avoids the hardcoded list of enemies for the nether fortress
feature, in favor of the (modifiable) list of spawns for the structure
feature as known by StructureSpawnManager.

sciwhiz12 committed:
Fix StructureSpawnManager to respect insideOnly

Fixes #8301

PickleDre commented:
sorry if this is a dumb question, but what files am i changing to fix this issue? cos i am struggling to find src/main/java/net/minecraftforge/common/world/StructureSpawnManager.java this anywhere

TheCurle reviewed (approved):
This is the piece i was missing when porting these functions.
Great catch.

Campbell314 commented:
Thanks for the fix!
